src/v2/manifest/mod.rs

```diff
@@ -238,3 +270,87 @@ fn to_mimes(v: &[&str]) -> Result<Vec<mime::Mime>> {
         .collect();
     Ok(res)
 }
+
+// Evaluate the `MediaTypes` from the the request header.
+fn evaluate_media_type(
+    content_type: Option<&http::header::HeaderValue>,
+    url: &Url,
+) -> Result<mediatypes::MediaTypes> {
+    let header_content_type = content_type
+        .map(|hv| hv.to_str())
+        .map(std::result::Result::unwrap_or_default);
+
+    let is_pulp_based = url.path().starts_with("/pulp/docker/v2");
+
+    match (header_content_type, is_pulp_based) {
+        (Some(header_value), false) => {
+            mediatypes::MediaTypes::from_str(header_value).map_err(Into::into)
+        }
+        (None, false) => Err(Error::from(
+            "no header_content_type given and no workaround to apply".to_string(),
+        )),
+        (Some(header_value), true) => {
+            // TODO: remove this workaround once Satellite returns a proper content-type here
+            match header_value {
+                "application/x-troff-man" => {
+                    trace!("Applying workaround for pulp-based registries, e.g. Satellite");
+                    mediatypes::MediaTypes::from_str(
+                        "application/vnd.docker.distribution.manifest.v1+prettyjws",
+                    )
+                    .map_err(Into::into)
+                }
+                _ => {
+                    debug!("Received content-type '{}' from pulp-based registry. Feeling lucky and trying to parse it...", header_value);
+                    mediatypes::MediaTypes::from_str(header_value).map_err(Into::into)
+                }
+            }
+        }
+        (None, true) => {
+            trace!("Applying workaround for pulp-based registries, e.g. Satellite");
+            mediatypes::MediaTypes::from_str(
+                "application/vnd.docker.distribution.manifest.v1+prettyjws",
+            )
+            .map_err(Into::into)
+        }
+    }
+}
+
+/// Umbrella type for common actions on the different manifest schema types
+#[derive(Debug)]
+pub enum Manifest {
+    S1Signed(manifest_schema1::ManifestSchema1Signed),
+    S2(manifest_schema2::ManifestSchema2),
+    ML(manifest_schema2::ManifestList),
+}
+
+impl Manifest {
+    /// List digests of all layers referenced by this manifest, if available.
+    ///
+    /// The returned layers list is ordered starting with the base image first.
+    pub fn layers_digests(&self) -> Result<Vec<String>> {
+        match self {
+            Manifest::S1Signed(m) => Ok(m.get_layers()),
+            Manifest::S2(m) => Ok(m.get_layers()),
+            // Manifest::ML(_) => TODO(steveeJ),
+            _ => Err(format!(
+                "Manifest {:?} doesn't support the 'layer_digests' method",
+                self
+            )
+            .into()),
+        }
+    }
+
+    /// The architecture of the image the manifest points to, if available.
+    pub fn architecture(&self) -> Result<String> {
```

Comment:
This should probably return a set, which could be empty (e.g. [`manifests` can be an empty-array][1] for the very-similar OCI indexes).  I don't care if we throw an error on that case or return an empty set, because OpenShift consumers will only care about images with architectures (there's no point to have OpenShift consumers look at an image that has no architecture).

[1]: https://github.com/opencontainers/image-spec/blame/v1.0.1/image-index.md#L26